In [ ]:
import pandas as pd 
pd.set_option('display.max_columns', None)

In [ ]:
# Load the dataset
app = pd.read_csv(r"C:\Users\Admin\Downloads\final_dataprep\dseb63_final_project_DP_dataset\dseb63_application_test.csv")

In [ ]:
# Check overall info
app.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 246009 entries, 0 to 246008
Columns: 123 entries, Unnamed: 0 to SK_ID_CURR
dtypes: float64(65), int64(42), object(16)
memory usage: 230.9+ MB


In [ ]:
# Drop the index column
app.drop(columns = 'Unnamed: 0', inplace = True)

In [ ]:
# Check for null values
null_counts = app.isnull().sum()
print(null_counts.iloc[:50])  # First 50 columns
print(null_counts.iloc[50:100])  # Next 50 columns

TARGET                              0
NAME_CONTRACT_TYPE                  0
CODE_GENDER                         0
FLAG_OWN_CAR                        0
FLAG_OWN_REALTY                     0
CNT_CHILDREN                        0
AMT_INCOME_TOTAL                    0
AMT_CREDIT                          0
AMT_ANNUITY                        10
AMT_GOODS_PRICE                   224
NAME_TYPE_SUITE                  1046
NAME_INCOME_TYPE                    0
NAME_EDUCATION_TYPE                 0
NAME_FAMILY_STATUS                  0
NAME_HOUSING_TYPE                   0
REGION_POPULATION_RELATIVE          0
DAYS_BIRTH                          0
DAYS_EMPLOYED                       0
DAYS_REGISTRATION                   0
DAYS_ID_PUBLISH                     0
OWN_CAR_AGE                    162188
FLAG_MOBIL                          0
FLAG_EMP_PHONE                      0
FLAG_WORK_PHONE                     0
FLAG_CONT_MOBILE                    0
FLAG_PHONE                          0
FLAG_EMAIL  

## Cleaning AMT_ANNUITY column

In [7]:
app[app['AMT_ANNUITY'].isna()]

TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  \
37959        0         Cash loans           M            N               N   
39959        0         Cash loans           F            N               N   
41211        0         Cash loans           F            N               N   
43978        0         Cash loans           F            N               N   
47940        0         Cash loans           M            Y               N   
60698        0         Cash loans           M            Y               N   
71475        0         Cash loans           F            Y               N   
99054        0         Cash loans           F            N               Y   
191407       0         Cash loans           F            N               Y   
193396       0         Cash loans           F            N               N   

        CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
37959              0          180000.0    450000.0          NaN   
39959              0           94500.0    450000.0          NaN   
41211              0          202500.0    539100.0          NaN   
43978              0          162000.0    296280.0          NaN   
47940              0          202500.0    360000.0          NaN   
60698              0          144000.0    219249.0          NaN   
71475              0           90000.0    157500.0          NaN   
99054              0          202500.0    929088.0          NaN   
191407             0          157500.0    792000.0          NaN   
193396             0          315000.0   1483231.5          NaN   

        AMT_GOODS_PRICE NAME_TYPE_SUITE      NAME_INCOME_TYPE  \
37959          450000.0   Unaccompanied  Commercial associate   
39959          450000.0   Unaccompanied               Working   
41211          450000.0   Unaccompanied               Working   
43978          225000.0   Unaccompanied         State servant   
47940          360000.0   Unaccompanied               Working   
60698          166500.0   Unaccompanied               Working   
71475          157500.0   Unaccompanied         State servant   
99054          720000.0   Unaccompanied               Working   
191407         792000.0          Family         State servant   
193396        1354500.0   Unaccompanied               Working   

                  NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  \
37959               Incomplete higher  Single / not married   
39959                 Lower secondary        Civil marriage   
41211   Secondary / secondary special               Married   
43978                Higher education               Married   
47940   Secondary / secondary special  Single / not married   
60698                Higher education  Single / not married   
71475   Secondary / secondary special               Married   
99054   Secondary / secondary special        Civil marriage   
191407  Secondary / secondary special               Married   
193396               Higher education               Married   

        NAME_HOUSING_TYPE  REGION_POPULATION_RELATIVE  DAYS_BIRTH  \
37959   House / apartment                    0.026392      -10668   
39959   House / apartment                    0.035792       -9027   
41211   House / apartment                    0.046220      -17376   
43978   House / apartment                    0.035792      -11329   
47940   House / apartment                    0.046220      -19762   
60698    Rented apartment                    0.022800      -20831   
71475   House / apartment                    0.015221      -12134   
99054   House / apartment                    0.019689      -13902   
191407  House / apartment                    0.026392      -17661   
193396  House / apartment                    0.072508      -15072   

        DAYS_EMPLOYED  DAYS_REGISTRATION  DAYS_ID_PUBLISH  OWN_CAR_AGE  \
37959           -2523            -4946.0            -3238          NaN   
39959           -1270            -3640.0             -741          NaN   
41211           -39

### In all rows that this column is NaN, the result shows that the contract type is “Cash loans”. Therefore, we fillna in this column with the AMT_CREDIT divided by the average months of all the customers requiring “Cash loans”

In [ ]:
# Filter "Cash loans" customers
cash_loans = app[app['NAME_CONTRACT_TYPE'] == 'Cash loans']

# Remove NaN in AMT_CREDIT and AMT_ANNUITY
cash_loans_filtered = cash_loans.dropna(subset=['AMT_CREDIT', 'AMT_ANNUITY'])

# Calculate the average months
cash_loans_filtered['MONTHS'] = cash_loans_filtered['AMT_CREDIT'] / cash_loans_filtered['AMT_ANNUITY']
average_months = cash_loans_filtered['MONTHS'].mean()

print(f"Số tháng trung bình cho hợp đồng 'Cash loans': {average_months:.2f}")

Số tháng trung bình cho hợp đồng 'Cash loans': 21.79


C:\Users\Admin\AppData\Local\Temp\ipykernel_20936\756088765.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cash_loans_filtered['MONTHS'] = cash_loans_filtered['AMT_CREDIT'] / cash_loans_filtered['AMT_ANNUITY']


In [ ]:
# Fill the NaN value in AMT_ANNUITY
avg_months = average_months 
app['AMT_ANNUITY'] = app.apply(
    lambda row: row['AMT_CREDIT'] / avg_months if pd.isnull(row['AMT_ANNUITY']) and row['NAME_CONTRACT_TYPE'] == 'Cash loans' else row['AMT_ANNUITY'], 
    axis=1
)

## Cleaning AMT_GOODS_PRICE column

In [10]:
app[app['AMT_GOODS_PRICE'].isna()]

TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  \
586          0    Revolving loans           F            N               Y   
4758         0    Revolving loans           F            N               N   
5139         0    Revolving loans           F            N               N   
5367         0    Revolving loans           F            N               N   
6137         0    Revolving loans           M            N               Y   
...        ...                ...         ...          ...             ...   
243716       0    Revolving loans           F            N               Y   
243797       0    Revolving loans           M            N               Y   
244636       0    Revolving loans           F            N               Y   
244871       0    Revolving loans           M            N               Y   
244989       0    Revolving loans           M            N               Y   

        CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
586                2           45000.0    135000.0       6750.0   
4758               0          157500.0    450000.0      22500.0   
5139               0           67500.0    202500.0      10125.0   
5367               1          121500.0    180000.0       9000.0   
6137               0          180000.0    450000.0      22500.0   
...              ...               ...         ...          ...   
243716             2           67500.0    202500.0      10125.0   
243797             1          112500.0    270000.0      13500.0   
244636             0          126000.0    270000.0      13500.0   
244871             0          135000.0    270000.0      13500.0   
244989             2           67500.0    202500.0      10125.0   

        AMT_GOODS_PRICE NAME_TYPE_SUITE      NAME_INCOME_TYPE  \
586                 NaN             NaN  Commercial associate   
4758                NaN             NaN               Working   
5139                NaN             NaN               Working   
5367                NaN             NaN               Working   
6137                NaN             NaN               Working   
...                 ...             ...                   ...   
243716              NaN             NaN               Working   
243797              NaN             NaN               Working   
244636              NaN             NaN               Working   
244871              NaN             NaN  Commercial associate   
244989              NaN             NaN               Working   

                  NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  \
586     Secondary / secondary special               Married   
4758    Secondary / secondary special               Married   
5139                 Higher education               Married   
5367    Secondary / secondary special        Civil marriage   
6137                 Higher education  Single / not married   
...                               ...                   ...   
243716               Higher education               Married   
243797  Secondary / secondary special               Married   
244636              Incomplete higher  Single / not married   
244871  Secondary / secondary special               Married   
244989               Higher education               Married   

          NAME_HOUSING_TYPE  REGION_POPULATION_RELATIVE  DAYS_BIRTH  \
586       House / apartment                    0.024610      -10072   
4758      House / apartment                    0.010006      -11993   
5139      House / apartment                    0.008474       -9727   
5367      House / apartment                    0.011657      -11079   
6137      House / apartment                    0.032561       -9986   
...                     ...                         ...         ...   
243716    House / apartment                    0.025164      -11313   
243797    House / apartment                    0.028663      -16870   
244636         With parents                    0.015221      -10440   
244871  Municipal 

### In all rows that this column is NaN, the result shows that the contract type is “Revolving loans”. Therefore, we fillna in this column with the median value where all the customers requiring “Revolving loans”

In [ ]:
# Median of AMT_GOODS_PRICE when NAME_CONTRACT_TYPE = "Revolving loans"
revolving_median = app.loc[app['NAME_CONTRACT_TYPE'] == 'Revolving loans', 'AMT_GOODS_PRICE'].median()

# Fillna by the median value calculated above
app.loc[(app['NAME_CONTRACT_TYPE'] == 'Revolving loans') & (app['AMT_GOODS_PRICE'].isna()), 'AMT_GOODS_PRICE'] = revolving_median

## Cleaning NAME_TYPE_SUITE column

In [ ]:
# Unique family status filtered that NAME_TYPE_SUITE is NaN
app[app['NAME_TYPE_SUITE'].isna()]['NAME_FAMILY_STATUS'].value_counts()

NAME_FAMILY_STATUS
Married                 648
Single / not married    191
Civil marriage           92
Widow                    60
Separated                54
Unknown                   1
Name: count, dtype: int64

In [ ]:
# Find the mode of NAME_TYPE_SUITE for each group of NAME_FAMILY_STATUS
mode_by_family_status = app.groupby('NAME_FAMILY_STATUS')['NAME_TYPE_SUITE'].agg(lambda x: x.mode()[0] if not x.mode().empty else 'Unknown')
mode_by_family_status

NAME_FAMILY_STATUS
Civil marriage          Unaccompanied
Married                 Unaccompanied
Separated               Unaccompanied
Single / not married    Unaccompanied
Unknown                       Unknown
Widow                   Unaccompanied
Name: NAME_TYPE_SUITE, dtype: object

In [ ]:
# Fill with the mode filtered for each NAME_FAMILY_STATUS group, as they have a domain relationship
app['NAME_TYPE_SUITE'] = app.apply(
    lambda row: mode_by_family_status[row['NAME_FAMILY_STATUS']] if pd.isna(row['NAME_TYPE_SUITE']) else row['NAME_TYPE_SUITE'],
    axis=1
)

## Cleaning OWN_CAR_AGE column

In [ ]:
# NaN rows in OWN_CAR_AGE
missing_car_age = app[app['OWN_CAR_AGE'].isna()]

# Check if these rows are the customers that don't own car (FLAG_OWN_CAR == 'N') or not
missing_car_age_no_car = missing_car_age[missing_car_age['FLAG_OWN_CAR'] == 'N']

# Check the result
print(f"Số lượng dòng thiếu OWN_CAR_AGE: {missing_car_age.shape[0]}")
print(f"Số lượng dòng thiếu OWN_CAR_AGE và không sở hữu xe: {missing_car_age_no_car.shape[0]}")

missing_car_age_no_car

Số lượng dòng thiếu OWN_CAR_AGE: 162188
Số lượng dòng thiếu OWN_CAR_AGE và không sở hữu xe: 162184


TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  \
0            0         Cash loans           F            N               N   
1            0         Cash loans           F            N               Y   
2            0         Cash loans           M            N               Y   
3            0         Cash loans           M            N               Y   
5            0         Cash loans           F            N               Y   
...        ...                ...         ...          ...             ...   
246004       0         Cash loans           M            N               N   
246005       0         Cash loans           F            N               Y   
246006       0         Cash loans           F            N               Y   
246007       1         Cash loans           F            N               Y   
246008       0         Cash loans           F            N               N   

        CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0                  0          270000.0   1293502.5      35698.5   
1                  0          135000.0    312682.5      29686.5   
2                  0          121500.0    513000.0      21865.5   
3                  0           99000.0    490495.5      27517.5   
5                  0          112500.0   1019610.0      33826.5   
...              ...               ...         ...          ...   
246004             0          157500.0    254700.0      27558.0   
246005             0           72000.0    269550.0      12001.5   
246006             0          153000.0    677664.0      29979.0   
246007             0          171000.0    370107.0      20205.0   
246008             0          157500.0    675000.0      49117.5   

        AMT_GOODS_PRICE  NAME_TYPE_SUITE      NAME_INCOME_TYPE  \
0             1129500.0           Family         State servant   
1              297000.0    Unaccompanied               Working   
2              513000.0    Unaccompanied               Working   
3              454500.0  Spouse, partner         State servant   
5              913500.0         Children             Pensioner   
...                 ...              ...                   ...   
246004         225000.0    Unaccompanied               Working   
246005         225000.0    Unaccompanied             Pensioner   
246006         585000.0    Unaccompanied               Working   
246007         319500.0    Unaccompanied  Commercial associate   
246008         675000.0    Unaccompanied  Commercial associate   

                  NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  \
0                    Higher education               Married   
1       Secondary / secondary special        Civil marriage   
2       Secondary / secondary special  Single / not married   
3       Secondary / secondary special               Married   
5       Secondary / secondary special               Married   
...                               ...                   ...   
246004  Secondary / secondary special             Separated   
246005  Secondary / secondary special                 Widow   
246006               Higher education             Separated   
246007  Secondary / secondary special               Married   
246008               Higher education               Married   

        NAME_HOUSING_TYPE  REGION_POPULATION_RELATIVE  DAYS_BIRTH  \
0       House / apartment                    0.003541      -16765   
1       House / apartment                    0.008019      -19005   
2       House / apartment                    0.028663      -19932   
3       House / apartment                    0.035792      -16941   
5       House / apartment                    0.018634      -20099   
...                   ...                         ...         ...   
246004       With parents                    0.032561       -9327   
246005  House / apartment                    0.025164      -20775   
246006  House / apartment                    0.005002      -14966   
246007  House / apartment 

In [16]:
app[(app['FLAG_OWN_CAR'] == 'Y') & (app['OWN_CAR_AGE'] == 0)]

TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  \
204          0         Cash loans           M            Y               N   
328          0         Cash loans           F            Y               Y   
423          0         Cash loans           F            Y               Y   
453          0         Cash loans           F            Y               Y   
646          0         Cash loans           M            Y               Y   
...        ...                ...         ...          ...             ...   
245564       0         Cash loans           M            Y               N   
245720       0         Cash loans           M            Y               Y   
245752       0         Cash loans           F            Y               Y   
245831       0         Cash loans           M            Y               Y   
245997       0         Cash loans           F            Y               Y   

        CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
204                0          202500.0    526491.0      26878.5   
328                0          225000.0   1345500.0      39469.5   
423                1          229500.0   1006920.0      72778.5   
453                2           76500.0    119925.0      13045.5   
646                0          337500.0    900000.0      26446.5   
...              ...               ...         ...          ...   
245564             0          315000.0    364896.0      23724.0   
245720             0          225000.0   1174090.5      46692.0   
245752             0          315000.0   1066500.0      31180.5   
245831             0          225000.0   1035000.0      30393.0   
245997             0          180000.0    270126.0      12028.5   

        AMT_GOODS_PRICE NAME_TYPE_SUITE      NAME_INCOME_TYPE  \
204            454500.0   Unaccompanied               Working   
328           1345500.0   Unaccompanied  Commercial associate   
423            900000.0   Unaccompanied               Working   
453            112500.0   Unaccompanied               Working   
646            900000.0   Unaccompanied  Commercial associate   
...                 ...             ...                   ...   
245564         315000.0   Unaccompanied               Working   
245720        1080000.0   Unaccompanied             Pensioner   
245752        1066500.0   Unaccompanied               Working   
245831        1035000.0   Unaccompanied               Working   
245997         193500.0   Unaccompanied               Working   

                  NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  \
204                  Higher education               Married   
328                  Higher education               Married   
423                  Higher education               Married   
453                  Higher education               Married   
646                  Higher education               Married   
...                               ...                   ...   
245564               Higher education  Single / not married   
245720  Secondary / secondary special               Married   
245752               Higher education               Married   
245831  Secondary / secondary special               Married   
245997  Secondary / secondary special               Married   

        NAME_HOUSING_TYPE  REGION_POPULATION_RELATIVE  DAYS_BIRTH  \
204     House / apartment                    0.022625      -14811   
328     House / apartment                    0.026392      -15601   
423     House / apartment                    0.008230       -9920   
453     House / apartment                    0.014464      -12463   
646     House / apartment                    0.022800      -21910   
...                   ...                         ...         ...   
245564  House / apartment                    0.019101       -9642   
245720  House / apartment                    0.019101      -22704   
245752  House / apartment                    0.019689      -12932   
245831  House / apartment             

### There are 162,188 NaN rows, of which 162,184 customers don’t own a car. So for those who own a car, we fillna with median, and for those who do not, we fillna with -1

In [ ]:
# Fill -1 for those who don't own car
app.loc[app['FLAG_OWN_CAR'] == 'N', 'OWN_CAR_AGE'] = -1

# Fill median for those who own car
median_car_age = app.loc[app['FLAG_OWN_CAR'] == 'Y', 'OWN_CAR_AGE'].median()
app.loc[(app['FLAG_OWN_CAR'] == 'Y') & (app['OWN_CAR_AGE'].isna()), 'OWN_CAR_AGE'] = median_car_age

## Cleaning OCCUPATION_TYPE column

In [ ]:
# Divide AMT_INCOME_TOTAL into groups (quantile)
app['INCOME_QUANTILE'] = pd.qcut(app['AMT_INCOME_TOTAL'], q=4, labels=['Low', 'Medium', 'High', 'Very High'])

In [ ]:
# Find the most popular value in each group
occupation_mode = app.groupby(['NAME_EDUCATION_TYPE', 'INCOME_QUANTILE'])['OCCUPATION_TYPE'].agg(lambda x: x.mode().iloc[0] if not x.mode().empty else None)
print(occupation_mode)

NAME_EDUCATION_TYPE            INCOME_QUANTILE
Academic degree                Low                   Managers
                               Medium                Managers
                               High                Core staff
                               Very High             Managers
Higher education               Low                 Core staff
                               Medium              Core staff
                               High                Core staff
                               Very High             Managers
Incomplete higher              Low                Sales staff
                               Medium                Laborers
                               High                  Laborers
                               Very High             Managers
Lower secondary                Low                   Laborers
                               Medium                Laborers
                               High                  Laborers
                       

C:\Users\Admin\AppData\Local\Temp\ipykernel_20936\1438009852.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  occupation_mode = app.groupby(['NAME_EDUCATION_TYPE', 'INCOME_QUANTILE'])['OCCUPATION_TYPE'].agg(lambda x: x.mode().iloc[0] if not x.mode().empty else None)


### Fill with the mode within groups based on NAME_EDUCATION_TYPE and INCOME_QUANTILE (low/medium/high/very high), as they have a domain relationship

In [20]:
def fill_occupation(row):
    if pd.isna(row['OCCUPATION_TYPE']):
        return occupation_mode.get((row['NAME_EDUCATION_TYPE'], row['INCOME_QUANTILE']), row['OCCUPATION_TYPE'])
    else:
        return row['OCCUPATION_TYPE']

app['OCCUPATION_TYPE'] = app.apply(fill_occupation, axis=1)

## Cleaning CNT_FAM_MEMBERS column

In [21]:
app[app['CNT_FAM_MEMBERS'].isna()]

TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  \
33550       0    Revolving loans           M            N               Y   

       CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
33550             0          450000.0    675000.0      33750.0   

       AMT_GOODS_PRICE NAME_TYPE_SUITE      NAME_INCOME_TYPE  \
33550         270000.0         Unknown  Commercial associate   

      NAME_EDUCATION_TYPE NAME_FAMILY_STATUS    NAME_HOUSING_TYPE  \
33550     Lower secondary            Unknown  Municipal apartment   

       REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  \
33550                    0.015221      -12396          -1161   

       DAYS_REGISTRATION  DAYS_ID_PUBLISH  OWN_CAR_AGE  FLAG_MOBIL  \
33550            -3265.0            -4489         -1.0           1   

       FLAG_EMP_PHONE  FLAG_WORK_PHONE  FLAG_CONT_MOBILE  FLAG_PHONE  \
33550               1                1                 1           1   

       FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  REGION_RATING_CLIENT  \
33550           0        Managers              NaN                     2   

       REGION_RATING_CLIENT_W_CITY WEEKDAY_APPR_PROCESS_START  \
33550                            2                   THURSDAY   

       HOUR_APPR_PROCESS_START  REG_REGION_NOT_LIVE_REGION  \
33550                       15                           0   

       REG_REGION_NOT_WORK_REGION  LIVE_REGION_NOT_WORK_REGION  \
33550                           1                            1   

       REG_CITY_NOT_LIVE_CITY  REG_CITY_NOT_WORK_CITY  \
33550                       0                       1   

       LIVE_CITY_NOT_WORK_CITY ORGANIZATION_TYPE  EXT_SOURCE_1  EXT_SOURCE_2  \
33550                        1         Insurance      0.628564      0.700618   

       EXT_SOURCE_3  APARTMENTS_AVG  BASEMENTAREA_AVG  \
33550           NaN             NaN               NaN   

       YEARS_BEGINEXPLUATATION_AVG  YEARS_BUILD_AVG  COMMONAREA_AVG  \
33550                          NaN              NaN             NaN   

       ELEVATORS_AVG  ENTRANCES_AVG  FLOORSMAX_AVG  FLOORSMIN_AVG  \
33550            NaN            NaN            NaN            NaN   

       LANDAREA_AVG  LIVINGAPARTMENTS_AVG  LIVINGAREA_AVG  \
33550           NaN                   NaN             NaN   

       NONLIVINGAPARTMENTS_AVG  NONLIVINGAREA_AVG  APARTMENTS_MODE  \
33550                      NaN                NaN              NaN   

       BASEMENTAREA_MODE  YEARS_BEGINEXPLUATATION_MODE  YEARS_BUILD_MODE  \
33550                NaN                           NaN               NaN   

       COMMONAREA_MODE  ELEVATORS_MODE  ENTRANCES_MODE  FLOORSMAX_MODE  \
33550              NaN             NaN             NaN             NaN   

       FLOORSMIN_MODE  LANDAREA_MODE  LIVINGAPARTMENTS_MODE  LIVINGAREA_MODE  \
33550             NaN            NaN                    NaN              NaN   

       NONLIVINGAPARTMENTS_MODE  NONLIVINGAREA_MODE  APARTMENTS_MEDI  \
33550                       NaN                 NaN              NaN   

       BASEMENTAREA_MEDI  YEARS_BEGINEXPLUATATION_MEDI  YEARS_BUILD_MEDI  \
33550                NaN                           NaN               NaN   

       COMMONAREA_MEDI  ELEVATORS_MEDI  ENTRANCES_MEDI  FLOORSMAX_MEDI  \
33550              NaN             NaN             NaN             NaN   

       FLOORSMIN_MEDI  LANDAREA_MEDI  LIVINGAPARTMENTS_MEDI  LIVINGAREA_MEDI  \
33550             NaN            NaN                    NaN              NaN   

       NONLIVINGAPARTMENTS_MEDI  NONLIVINGAREA_MEDI FONDKAPREMONT_MODE  \
33550                       NaN                 NaN                NaN   

      HOUSETYPE_MODE  TOTALAREA_MODE WALLSMATERIAL_MODE EMERGENCYSTATE_MODE  \
33550            NaN             NaN                NaN                 NaN   

       OBS_30_CNT_SOCIAL_CIRCLE  DEF_30_CNT_SOCIAL_CIRCLE  \
33550                       3.0                       0.0   

       OBS_60_CNT_SOCIAL_CIRCLE  DEF_60_CNT_SOCIAL_CIRC

### Only 1 NaN value in a customer whose family status is unknown, therefore fillna with 0

In [22]:
app['CNT_FAM_MEMBERS'].fillna(value= 0, inplace= True)

C:\Users\Admin\AppData\Local\Temp\ipykernel_20936\126323430.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  app['CNT_FAM_MEMBERS'].fillna(value= 0, inplace= True)


## Cleaning EXT_SOURCE columns

In [23]:
app[['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']].describe()

EXT_SOURCE_1  EXT_SOURCE_2   EXT_SOURCE_3
count  107484.000000  2.454800e+05  197185.000000
mean        0.502462  5.143854e-01       0.511034
std         0.210813  1.910533e-01       0.194837
min         0.014568  8.173617e-08       0.000527
25%         0.334376  3.922228e-01       0.370650
50%         0.506427  5.658916e-01       0.537070
75%         0.675074  6.637452e-01       0.669057
max         0.951624  8.549997e-01       0.896010

In [ ]:
# Fillna with the median as this comes from external information
for col in ['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']:
    app[col].fillna(app[col].median(), inplace=True)

C:\Users\Admin\AppData\Local\Temp\ipykernel_20936\2578508374.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  app[col].fillna(app[col].median(), inplace=True)
C:\Users\Admin\AppData\Local\Temp\ipykernel_20936\2578508374.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For exa

## Cleaning the remaining columns

In [ ]:
# Find the other missing columns
missing_cols = app.columns[app.isna().any()].tolist()

# For other columns, fillna with the median where columns are numerical, otherwise fillna with the mode
for col in missing_cols:
    if app[col].dtype == 'object': 
        app[col] = app[col].fillna(app[col].mode()[0])
    else:  
        app[col] = app[col].fillna(app[col].median())  

In [ ]:
# Export cleaned dataset
file_path = r'C:\Users\Admin\Desktop\Final data prep\cleaned_data\application_test.csv'
app.to_csv(file_path, index=False)

print(f"CSV file has been saved to {file_path}")

CSV file has been saved to C:\Users\Admin\Desktop\Final data prep\cleaned_data\application_train.csv
